compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPU2DScalableTo1DScalable.cpp

```diff
@@ -0,0 +1,193 @@
+    // there's no other targets that support > 1D scalability).
+    auto targetAttr = IREE::HAL::ExecutableTargetAttr::lookup(op);
+    bool isArmSME = assumeArmSME || hasSMEFeature(targetAttr);
+    if (!isArmSME || opKnownToSupport2DScalableVectorizationWithArmSME(op))
+      return failure();
+
+    auto loweringConfigAttr = getLoweringConfig(op);
+    if (!loweringConfigAttr)
+      return failure();
+
+    auto tileSizes = loweringConfigAttr.getTileSizeVals();
+    auto scalableFlags = loweringConfigAttr.getScalableTileFlagVals();
+
+    // 1. Drop scalable dimensions from leading tiling levels first (and leading
+    // dimensions within tiling levels). This works out as dropping scalability
+    // from leading dimensions of the vector type.
+    int64_t firstTilingLevelWithScalableDims = -1;
+    int64_t numScalableDims = 0;
+    for (auto [level, levelScalableFlags] : llvm::enumerate(scalableFlags)) {
+      numScalableDims += llvm::count(levelScalableFlags, true);
+      if (numScalableDims > 0 && firstTilingLevelWithScalableDims == -1)
+        firstTilingLevelWithScalableDims = level;
+    }
+
+    if (numScalableDims <= 1)
+      return failure();
+
+    auto levelTileSizes = tileSizes[firstTilingLevelWithScalableDims];
+    auto levelScalableFlags = scalableFlags[firstTilingLevelWithScalableDims];
+
+    SmallVector<int64_t> loopTileSizes;
+    SmallVector<bool> newScalableFlags;
+    for (auto [flag, size] : llvm::zip(levelScalableFlags, levelTileSizes)) {
+      if (flag && numScalableDims >= 2) {
+        --numScalableDims;
+        loopTileSizes.push_back(size);
+        newScalableFlags.push_back(false);
+      } else {
+        loopTileSizes.push_back(0);
+        newScalableFlags.push_back(flag);
+      }
+    }
+
+    // 2. Re-tile the operation with some scalability dropped. This introduces
+    // loops for previously scalable vector/tile sizes.
+    scf::SCFTilingOptions options{};
+    setSCFTileSizes(options, op, loopTileSizes, {});
+    auto tilingResult =
+        scf::tileUsingSCF(rewriter, cast<TilingInterface>(op), options);
+    if (failed(tilingResult))
+      return failure();
+
+    // 3. Update the lowering config of the new tiled operations.
+    scalableFlags[firstTilingLevelWithScalableDims] = newScalableFlags;
+    auto newLoweringConfig = IREE::Codegen::LoweringConfigAttr::get(
+        getContext(), tileSizes, scalableFlags);
+    for (auto *newOp : tilingResult->tiledOps) {
+      if (isa<TilingInterface>(newOp))
+        setLoweringConfig(newOp, newLoweringConfig);
+    }
+
+    rewriter.replaceOp(op, tilingResult->replacements);
+    return success();
+  };
+
+private:
+  bool assumeArmSME{false};
+};
+
+void LLVMCPU2DScalableTo1DScalablePass::runOnOperation() {
+  RewritePatternSet patterns(&getContext());
+  patterns.add<DropUnsupportedScalableDimsFromTilingInterfaceOps>(
+      patterns.getContext(), assumeArmSME);
+
+  if (mlir::failed(mlir::applyPatternsAndFoldGreedily(getOperation(),
```

Comment:
Doing tiling with patterns has been tricky. It might be easier to just walk the IR and get the ops you want to transform and apply the tiling transformation to it directly (i..e not use the pattern rewriter mechanism)

Reply:
What's the issue? I didn't notice a problem while working on this :sweat_smile:

Reply:
Pattern rewriting is not a good fit for tiling. Learning from the past development, we usually need a filter/attribute to represent the state of the op. E.g., which op should be tiled, and which op is already tiled. It introduces the "state" semantics in the IR. Otherwise, it falls in an infinite pattern rewriting loop. If we take a close look at what should be happening, it is always

1. Collect the target operations.
2. Apply the tiling.
3. End.

In your use case, it is fine because the matcher is "whether the op has a lowering_config or not". IMO, we want to get rid of filter-based tiling patterns. Having a consistent way to apply tiling in IREE could prevent such patterns created again. So I'd vote to not use patterns here.

Reply:
> it is fine because the matcher is "whether the op has a lowering_config or not"

Sorry, I misread the code. So you're recursively to undo scalable tiling? I think in this case, we can use the `getVectorTileSizes` method from `TilingConfig`. Because scalable flag is only meaningful to vector level of tiling. (we should probably add it to the pipeline verifier). And we can easily extend this from `2D->1D` to `nD->1D`.

side note: I'm thinking if we can carry such tiling information to lowering_config, so we don't need to reset lowering_config. Updating configs during lowering pipeline is not a good sign because users do not expect that. One of examples that I can think of is about tuning. In tuning, we can preset configs, compile and benchmark. This could lead users to a weird state. Because the codegen behavior could be different from what users expect.  I don't have a solution yet, though.

Reply:
I'm aware of `getVectorTileSizes()` but I didn't find it to be very helpful here. The vector sizes can be split between the parallel, reduction, and inner parallel tiling levels, and that information is not returned from that function. So since I want to update the lowering config attribute, I found it easier just to look at it directly.

Reply:
I think `getVectorTileSizes` returns a joint list from parallel, reduction and inner parallel tile sizes. That gives you all you need right? E.g., the lowering_config on a matmul op could be `[[d_x, d_y, 0], [[4], [4], 0], [0, 0, 1]]`; the `getVectorTileSizes()` returns `[[4], [4], 1]` in this case.

Reply:
I'm not sure I follow. I want to update the `lowering_config`, having the joint list i.e. `[[4], [4], 1]` does not help because I have to map that back to the `lowering_config` tiling levels (which is not a 1-to-1 mapping, there's multiple lowering configs which could result in the same `getVectorTileSizes()`).

Reply:
Sorry, I missed the update part. In this case, I think we want to add a method to update vector tile sizes in TilingConfig. The main concern is about updating lowering_config. Updating lowering_config after KernelDispatch stage is a red flag to me. If we can't get rid of it, let's at least access and update it through a defined class/struct (which is TilingConfig in this case).

Reply:
The `TilingConfig` does not provide any way to update tile sizes, it simply provides methods for reading the `lowering_config`.

Reply:
I meant it is a missing feature if we want to update lowering_config. We need to implement it. I can try to implement it if you're more comfortable about it.

Reply:
I can implement it, but it'll have to re-create the `lowering_config` attribute (as I believe the actual attribute data is immutable).

Reply:
i.e. Something like:
```c++
LoweringConfigAttr TilingConfig::withNewVectorSizes(ArrayRef<int64_t> sizes, ArrayRef<bool> scalableFlags);
```

Reply:
This sounds good to me, thanks!

Reply:
Updated to avoid using patterns, and use the suggested `withNewVectorSizes()` helper.